File: stereography.py
from vpython import sphere, vector, curve, canvas, color, rate, scene, cross, dot, norm

# Create a canvas
scene = canvas(title='Stereographic Projection', width=800, height=600, center=vector(0, 0, 0), background=color.white)

# Create a sphere (representing the unit sphere)
R = 1  # Radius of the sphere
center=vector(0,0,R)
earth = sphere(pos=center, radius=R, color=color.blue, opacity=0.5)

# Lists to store drawn curves
drawn_curves_sphere = []
drawn_curves_plane = []

# Flag to indicate if the mouse is being pressed
is_drawing = False
current_curve_sphere = None
current_curve_plane = None

# Function for reverse stereographic projection
def stereographic_projection(x, y, z):
    denom = 2*R - z
    return (2*R * x / denom, 2*R * y / denom)

# Event handler for mouse down
def on_mousedown(evt):
    global is_drawing, current_curve_sphere, current_curve_plane
    is_drawing = True
    current_curve_sphere = curve(color=color.red)
    current_curve_plane = curve(color=color.green)
    drawn_curves_sphere.append(current_curve_sphere)
    drawn_curves_plane.append(current_curve_plane)
    add_point(evt.pos)

# Event handler for mouse move
def on_mousemove(evt):
    if is_drawing:
        add_point(evt.pos)

# Event handler for mouse up
def on_mouseup(evt):
    global is_drawing
    is_drawing = False

#Function to find if point is inside sphere
def point_inside_sphere(pos):
    distance = center - pos
    if distance.mag < R:
        return 1
    else:
        return -1
#Scaling factor to scale viwing direction appropriately to match surface of sphere (needs corrections)
def scale_to_sphere_factor(sign, direction, pos):
    p2=(pos-center).mag**2
    d2=direction.mag**2
    x2 = (p2 + sign*(R**2))/d2
    #print("x^2 = ",x2)
    x = pow(x2,0.5)
    return x


# Function to add point to the current curves
def add_point(pos):
    #print(pos)
    normal = norm(scene.camera.axis)
    shift = - normal * (abs(dot(normal,(center - pos))))
    shifted_pos = pos + shift
    
    scaling_factor_to_sphere = pow((pow(R,2) - pow((shifted_pos - center).mag,2)),0.5)
    sphere_point = shifted_pos - (normal * scaling_factor_to_sphere)
    print("shift = ",shift) ; print("center = ",center); print("pos = ",pos); print("normal = ",normal); print("sphere_point = ",sphere_point)
    current_curve_sphere.append(sphere_point)
    
    # Project this point onto the plane using stereographic projection
    proj_x, proj_y = stereographic_projection(sphere_point.x, sphere_point.y, sphere_point.z)
    proj_point = vector(proj_x, proj_y, 0)
    
    # Add the projected point to the plane curve
    current_curve_plane.append(proj_point)

# Event handler for key presses
def on_keydown(evt):
    global drawn_curves_sphere, drawn_curves_plane
    if evt.key == 'r':
        # Remove all drawn curves on the sphere and the plane
        for curve_obj in drawn_curves_sphere:
            curve_obj.visible = False  # Hide the curve on the sphere
        for curve_obj in drawn_curves_plane:
            curve_obj.visible = False  # Hide the curve on the plane
        drawn_curves_sphere.clear()  # Clear the list for sphere curves
        drawn_curves_plane.clear()  # Clear the list for plane curves

# Bind the mouse and key events to their handlers
scene.bind('mousedown', on_mousedown)
scene.bind('mousemove', on_mousemove)
scene.bind('mouseup', on_mouseup)
scene.bind('keydown', on_keydown)

# Draw a grid to represent the projection plane
plane = curve(vector(-2, -2, 0), vector(2, -2, 0), vector(2, 2, 0), vector(-2, 2, 0), vector(-2, -2, 0), color=color.black)
while True:
    rate(10)  # Run the loop at a reasonable rate (10 frames per second)

File: stereography.cpp
#include <GL/glut.h>
#include <glm/glm.hpp>
#include <glm/gtc/matrix_transform.hpp>
#include <glm/gtx/norm.hpp>
#include <vector>
#include <iostream>

// Variables
float R = 1.0f; // Radius of the sphere
glm::vec3 center(0.0f, 0.0f, R);
std::vector<std::vector<glm::vec3>> drawn_curves_sphere;
std::vector<std::vector<glm::vec3>> drawn_curves_plane;
bool is_drawing = false;
std::vector<glm::vec3> current_curve_sphere;
std::vector<glm::vec3> current_curve_plane;

// Function to perform stereographic projection
glm::vec2 stereographic_projection(float x, float y, float z) {
    float denom = 2 * R - z;
    return glm::vec2(2 * R * x / denom, 2 * R * y / denom);
}

// Function to check if a point is inside the sphere
int point_inside_sphere(glm::vec3 pos) {
    float distance = glm::length(center - pos);
    return (distance < R) ? 1 : -1;
}

// Function to calculate the scaling factor to the sphere
float scale_to_sphere_factor(int sign, glm::vec3 direction, glm::vec3 pos) {
    float p2 = glm::length2(pos - center);
    float d2 = glm::length2(direction);
    float x2 = (p2 + sign * (R * R)) / d2;
    return sqrt(x2);
}

// Function to erase drawn curves
void erase() {
    drawn_curves_sphere.clear();
    drawn_curves_plane.clear();
}

// Function to add a point to the current curve
void add_point(glm::vec3 pos) {
    glm::vec3 normal = glm::normalize(glm::vec3(0.0f, 0.0f, 1.0f)); // Assume normal in Z direction
    glm::vec3 shift = -normal * glm::dot(normal, (center - pos));
    glm::vec3 shifted_pos = pos + shift;
    float temp = R * R - glm::length2(shifted_pos - center);
    if (temp > 0) {
        float scaling_factor_to_sphere = sqrt(temp);
        glm::vec3 sphere_point = shifted_pos - (normal * scaling_factor_to_sphere);
        current_curve_sphere.push_back(sphere_point);

        glm::vec2 proj = stereographic_projection(sphere_point.x, sphere_point.y, sphere_point.z);
        glm::vec3 proj_point(proj.x, proj.y, 0.0f);

        current_curve_plane.push_back(proj_point);
    }
}

// Mouse and Keyboard Event Handling
void on_mouse(int button, int state, int x, int y) {
    if (button == GLUT_LEFT_BUTTON && state == GLUT_DOWN) {
        is_drawing = true;
        current_curve_sphere.clear();
        current_curve_plane.clear();
        drawn_curves_sphere.push_back(current_curve_sphere);
        drawn_curves_plane.push_back(current_curve_plane);
    }
    if (button == GLUT_LEFT_BUTTON && state == GLUT_UP) {
        is_drawing = false;
    }
}

void on_motion(int x, int y) {
    if (is_drawing) {
        glm::vec3 pos = glm::vec3(x / 100.0f, y / 100.0f, 0.0f);
        add_point(pos);
    }
}

void on_keyboard(unsigned char key, int x, int y) {
    if (key == 'r') {
        erase();
    }
}

// Function to draw the scene
void draw_scene() {
    glClear(GL_COLOR_BUFFER_BIT | GL_DEPTH_BUFFER_BIT);

    glColor3f(0.0f, 0.0f, 1.0f);
    glutSolidSphere(R, 50, 50); // Draw the sphere

    // Draw the curves on the sphere
    glColor3f(1.0f, 0.0f, 0.0f);
    for (auto &curve : drawn_curves_sphere) {
        glBegin(GL_LINE_STRIP);
        for (auto &point : curve) {
            glVertex3f(point.x, point.y, point.z);
        }
        glEnd();
    }

    // Draw the curves on the plane
    glColor3f(0.0f, 1.0f, 0.0f);
    for (auto &curve : drawn_curves_plane) {
        glBegin(GL_LINE_STRIP);
        for (auto &point : curve) {
            glVertex3f(point.x, point.y, point.z);
        }
        glEnd();
    }

    glutSwapBuffers();
}

void setup() {
    glClearColor(1.0f, 1.0f, 1.0f, 1.0f);
    glEnable(GL_DEPTH_TEST);
}

int main(int argc, char **argv) {
    glutInit(&argc, argv);
    glutInitDisplayMode(GLUT_DOUBLE | GLUT_RGB | GLUT_DEPTH);
    glutInitWindowSize(800, 600);
    glutCreateWindow("Stereographic Projection");

    setup();

    glutDisplayFunc(draw_scene);
    glutIdleFunc(draw_scene);
    glutMouseFunc(on_mouse);
    glutMotionFunc(on_motion);
    glutKeyboardFunc(on_keyboard);

    glutMainLoop();
    return 0;
}
